File: 1081-smallest-subsequence-of-distinct-characters/1081-smallest-subsequence-of-distinct-characters.cpp
class Solution {
public:
    string smallestSubsequence(string s) {
        int n=s.length();
        string result;
        vector<bool> token(26, false);
        vector<int> lastindex(26);
        for(int i=0; i<n; i++){
            lastindex[s[i] - 'a'] = i;
        }
        for(int i=0; i<n; i++){
            char ch=s[i];
            int idx=ch-'a';
            if(token[idx])
                continue;
             while (!result.empty() && result.back() > ch && lastindex[result.back() - 'a'] > i) {
                token[result.back() - 'a'] = false;
                result.pop_back();
            }
            result.push_back(ch);
            token[idx]=true;
        }
        return result;
    }
};

File: 1081-smallest-subsequence-of-distinct-characters/1081-smallest-subsequence-of-distinct-characters.java
class Solution {
    public String smallestSubsequence(String s) {
        int n = s.length();
        StringBuilder result = new StringBuilder();
        boolean[] visited = new boolean[26];
        int[] lastIndex = new int[26];
        for (int i = 0; i < n; i++) {
            lastIndex[s.charAt(i) - 'a'] = i;
        }
        for (int i = 0; i < n; i++) {
            char ch = s.charAt(i);
            int idx = ch - 'a';
            if (visited[idx])
                continue;
            while (result.length() > 0 && result.charAt(result.length() - 1) > ch && lastIndex[result.charAt(result.length() - 1) - 'a'] > i) {
                visited[result.charAt(result.length() - 1) - 'a'] = false;
                result.deleteCharAt(result.length() - 1);
            }
            result.append(ch);
            visited[idx] = true;
        }
        return result.toString();
    }
}